File: scripts/export_wcvp_native_ranges.R
#!/usr/bin/env Rscript

# Export exact-name accepted-species native range sizes from the WCVP data
# package plus the WGSRPD level-3 geometry used by WCVP/POWO. This is a
# corroboration layer for endemism, not a replacement taxonomy for the island
# master.

suppressPackageStartupMessages({
  library(rWCVP)
  library(rWCVPdata)
  library(dplyr)
  library(readr)
  library(sf)
  library(jsonlite)
})

args <- commandArgs(trailingOnly = TRUE)
if (length(args) < 2) {
  stop("Usage: export_wcvp_native_ranges.R ISLAND_TAXA_CSV OUTPUT_DIR")
}
input_path <- args[[1]]
out_dir <- args[[2]]
dir.create(out_dir, recursive = TRUE, showWarnings = FALSE)

target <- readr::read_csv(input_path, show_col_types = FALSE) %>%
  transmute(accepted_species = trimws(as.character(accepted_species))) %>%
  filter(accepted_species != "") %>%
  distinct()

names_tbl <- rWCVPdata::wcvp_names
dist_tbl <- rWCVPdata::wcvp_distributions

required_names <- c("plant_name_id", "taxon_name", "taxon_rank", "taxon_status")
missing_names <- setdiff(required_names, names(names_tbl))
if (length(missing_names)) stop(paste("WCVP names missing:", paste(missing_names, collapse=", ")))
required_dist <- c("plant_name_id", "area_code_l3", "introduced", "extinct", "location_doubtful")
missing_dist <- setdiff(required_dist, names(dist_tbl))
if (length(missing_dist)) stop(paste("WCVP distributions missing:", paste(missing_dist, collapse=", ")))

accepted <- names_tbl %>%
  mutate(
    taxon_name = trimws(as.character(taxon_name)),
    taxon_rank = tolower(as.character(taxon_rank)),
    taxon_status = tolower(as.character(taxon_status))
  ) %>%
  filter(taxon_rank == "species", taxon_status == "accepted") %>%
  inner_join(target, by = c("taxon_name" = "accepted_species")) %>%
  select(accepted_species = taxon_name, plant_name_id) %>%
  distinct()

native <- dist_tbl %>%
  inner_join(accepted, by = "plant_name_id") %>%
  mutate(
    introduced = as.numeric(introduced),
    extinct = as.numeric(extinct),
    location_doubtful = as.numeric(location_doubtful),
    area_code_l3 = trimws(as.character(area_code_l3))
  ) %>%
  filter(introduced == 0, extinct == 0, location_doubtful == 0, area_code_l3 != "") %>%
  distinct(accepted_species, plant_name_id, area_code_l3)

summary <- native %>%
  arrange(accepted_species, area_code_l3) %>%
  group_by(accepted_species, plant_name_id) %>%
  summarise(
    n_native_l3 = n_distinct(area_code_l3),
    native_l3_codes = paste(sort(unique(area_code_l3)), collapse = "|"),
    .groups = "drop"
  )

all_matched <- accepted %>%
  left_join(summary, by = c("accepted_species", "plant_name_id")) %>%
  mutate(
    n_native_l3 = ifelse(is.na(n_native_l3), 0L, as.integer(n_native_l3)),
    native_l3_codes = ifelse(is.na(native_l3_codes), "", native_l3_codes)
  )

readr::write_csv(all_matched, file.path(out_dir, "wcvp_native_range_summary.csv.gz"))
st_write(rWCVP::wgsrpd3, file.path(out_dir, "wgsrpd_level3.gpkg"), delete_dsn = TRUE, quiet = TRUE)

manifest <- list(
  contract = "wcvp_native_range_exact_name_v1",
  n_target_species = nrow(target),
  n_exact_accepted_species = nrow(accepted),
  n_species_with_native_range = sum(all_matched$n_native_l3 > 0),
  n_single_native_l3_species = sum(all_matched$n_native_l3 == 1),
  endemism_policy = paste(
    "A single WCVP native TDWG-L3 region is corroboration only; final endemic",
    "classification also requires an independent GIFT source-reported endemic claim",
    "and focal-island membership in the same TDWG-L3 region."
  )
)
jsonlite::write_json(manifest, file.path(out_dir, "wcvp_native_range_manifest.json"), auto_unbox=TRUE, pretty=TRUE)
print(jsonlite::toJSON(manifest, auto_unbox=TRUE, pretty=TRUE))

File: scripts/export_wcvp_native_ranges.py
#!/usr/bin/env python3
"""Export exact-name WCVP native TDWG-L3 ranges from the official bulk ZIP."""

from __future__ import annotations

import argparse
import json
import zipfile
from pathlib import Path

import pandas as pd


def _text(series: pd.Series) -> pd.Series:
    return series.fillna("").astype(str).str.strip()


def export_ranges(island_taxa_csv: Path, wcvp_zip: Path, output_dir: Path) -> dict[str, object]:
    target = pd.read_csv(island_taxa_csv, usecols=["accepted_species"])
    target["accepted_species"] = _text(target["accepted_species"])
    target_names = set(target.loc[target["accepted_species"].ne(""), "accepted_species"])

    output_dir.mkdir(parents=True, exist_ok=True)
    with zipfile.ZipFile(wcvp_zip) as archive:
        if "wcvp_names.csv" not in archive.namelist() or "wcvp_distribution.csv" not in archive.namelist():
            raise RuntimeError("official WCVP ZIP lacks expected names/distribution tables")

        matched_chunks: list[pd.DataFrame] = []
        with archive.open("wcvp_names.csv") as handle:
            for chunk in pd.read_csv(
                handle,
                sep="|",
                usecols=["plant_name_id", "taxon_rank", "taxon_status", "taxon_name"],
                dtype={"plant_name_id": "Int64", "taxon_rank": str, "taxon_status": str, "taxon_name": str},
                chunksize=250_000,
                low_memory=False,
            ):
                chunk["taxon_name"] = _text(chunk["taxon_name"])
                keep = (
                    _text(chunk["taxon_rank"]).str.casefold().eq("species")
                    & _text(chunk["taxon_status"]).str.casefold().eq("accepted")
                    & chunk["taxon_name"].isin(target_names)
                )
                selected = chunk.loc[keep, ["plant_name_id", "taxon_name"]].copy()
                if not selected.empty:
                    selected = selected.rename(columns={"taxon_name": "accepted_species"})
                    matched_chunks.append(selected)
        matched = (
            pd.concat(matched_chunks, ignore_index=True).dropna(subset=["plant_name_id"]).drop_duplicates()
            if matched_chunks
            else pd.DataFrame(columns=["plant_name_id", "accepted_species"])
        )
        if matched.empty:
            raise RuntimeError("no fixed-master species exactly matched accepted WCVP species")
        matched["plant_name_id"] = matched["plant_name_id"].astype("int64")
        ids = set(matched["plant_name_id"].tolist())
        id_to_species = matched.set_index("plant_name_id")["accepted_species"].to_dict()

        distribution_chunks: list[pd.DataFrame] = []
        with archive.open("wcvp_distribution.csv") as handle:
            for chunk in pd.read_csv(
                handle,
                sep="|",
                usecols=[
                    "plant_name_id",
                    "area_code_l3",
                    "introduced",
                    "extinct",
                    "location_doubtful",
                ],
                dtype={"plant_name_id": "Int64", "area_code_l3": str},
                chunksize=500_000,
                low_memory=False,
            ):
                chunk = chunk.loc[chunk["plant_name_id"].isin(ids)].copy()
                if chunk.empty:
                    continue
                for column in ("introduced", "extinct", "location_doubtful"):
                    chunk[column] = pd.to_numeric(chunk[column], errors="coerce")
                chunk["area_code_l3"] = _text(chunk["area_code_l3"])
                chunk = chunk.loc[
                    chunk["introduced"].eq(0)
                    & chunk["extinct"].eq(0)
                    & chunk["location_doubtful"].eq(0)
                    & chunk["area_code_l3"].ne("")
                ].copy()
                if not chunk.empty:
                    chunk["accepted_species"] = chunk["plant_name_id"].map(id_to_species)
                    distribution_chunks.append(
                        chunk[["accepted_species", "plant_name_id", "area_code_l3"]]
                    )

    native = (
        pd.concat(distribution_chunks, ignore_index=True).drop_duplicates()
        if distribution_chunks
        else pd.DataFrame(columns=["accepted_species", "plant_name_id", "area_code_l3"])
    )
    if native.empty:
        summary = matched.copy()
        summary["n_native_l3"] = 0
        summary["native_l3_codes"] = ""
    else:
        grouped = (
            native.groupby(["accepted_species", "plant_name_id"], as_index=False)
            .agg(
                n_native_l3=("area_code_l3", "nunique"),
                native_l3_codes=("area_code_l3", lambda x: "|".join(sorted(set(x)))),
            )
        )
        summary = matched.merge(
            grouped,
            on=["accepted_species", "plant_name_id"],
            how="left",
            validate="one_to_one",
        )
        summary["n_native_l3"] = summary["n_native_l3"].fillna(0).astype(int)
        summary["native_l3_codes"] = summary["native_l3_codes"].fillna("")

    summary.sort_values("accepted_species").to_csv(
        output_dir / "wcvp_native_range_summary.csv.gz", index=False
    )
    manifest = {
        "contract": "official_wcvp_bulk_exact_name_native_range_v1",
        "n_target_species": len(target_names),
        "n_exact_accepted_species": int(summary["accepted_species"].nunique()),
        "n_species_with_native_range": int(summary["n_native_l3"].gt(0).sum()),
        "n_single_native_l3_species": int(summary["n_native_l3"].eq(1).sum()),
        "source_files": ["wcvp_names.csv", "wcvp_distribution.csv"],
        "distribution_filters": {
            "introduced": 0,
            "extinct": 0,
            "location_doubtful": 0,
        },
        "policy": (
            "Exact accepted species only. A single native TDWG-L3 range is corroboration "
            "only; final endemic status also requires an independent GIFT endemic claim "
            "and focal-island membership in the same TDWG region."
        ),
    }
    (output_dir / "wcvp_native_range_manifest.json").write_text(
        json.dumps(manifest, indent=2) + "\n", encoding="utf-8"
    )
    return manifest


def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument("island_taxa_csv", type=Path)
    parser.add_argument("wcvp_zip", type=Path)
    parser.add_argument("output_dir", type=Path)
    args = parser.parse_args()
    print(json.dumps(export_ranges(args.island_taxa_csv, args.wcvp_zip, args.output_dir), indent=2))


if __name__ == "__main__":
    main()
